include/alpaka/dev/DevOmp5.hpp

```diff
@@ -0,0 +1,365 @@
+
+            ALPAKA_FN_HOST auto getAllQueues() const
+            -> std::vector<std::shared_ptr<queue::IGenericThreadsQueue<DevOmp5>>>
+            {
+                return m_spDevOmp5Impl->getAllExistingQueues();
+            }
+
+            //-----------------------------------------------------------------------------
+            //! Registers the given queue on this device.
+            //! NOTE: Every queue has to be registered for correct functionality of device wait operations!
+            ALPAKA_FN_HOST auto registerQueue(std::shared_ptr<queue::IGenericThreadsQueue<DevOmp5>> spQueue) const
+            -> void
+            {
+                m_spDevOmp5Impl->registerQueue(spQueue);
+            }
+
+        public:
+            std::shared_ptr<omp5::detail::DevOmp5Impl> m_spDevOmp5Impl;
+        };
+    }
+
+    namespace dev
+    {
+        namespace traits
+        {
+            //#############################################################################
+            //! The CUDA RT device name get trait specialization.
+            template<>
+            struct GetName<
+                dev::DevOmp5>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getName(
+                    dev::DevOmp5 const &)
+                -> std::string
+                {
+                    return std::string("OMP5 target");
+                }
+            };
+
+            //#############################################################################
+            //! The CUDA RT device available memory get trait specialization.
+            template<>
+            struct GetMemBytes<
+                dev::DevOmp5>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getMemBytes(
+                    dev::DevOmp5 const & dev)
+                -> std::size_t
+                {
+                    alpaka::ignore_unused(dev); //! \TODO
+                    // std::size_t freeInternal(0u);
+                    std::size_t totalInternal(6ull<<30); //! \TODO
+
+                    return totalInternal;
+                }
+            };
+
+            //#############################################################################
+            //! The CUDA RT device free memory get trait specialization.
+            template<>
+            struct GetFreeMemBytes<
+                dev::DevOmp5>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getFreeMemBytes(
+                    dev::DevOmp5 const & dev)
+                -> std::size_t
+                {
+                    alpaka::ignore_unused(dev); //! \todo query device
+                    std::size_t freeInternal((6ull<<30));
+                    // std::size_t totalInternal(0u);
```

Comment:
Any ideas here and above?

Reply:
The OpenMP standard does not offer a way to query this. AFAIK, there is also no way to figure what type of device the target is, so we could check if we are targeting the CPU, then check the host memory, or if the we are targeting the GPU and then ask the CUDA/HIP runtime (which would be problematinc any because it would become a dependency). Thought of a workaround like this was the reason for the `\todo`. I think this can just not be supported. What is the alpaka policy to report, that this number is not available?

Reply:
Any non-zero value might be too much on a given hardware.
Only pessimistic answers can be valid answers. As we do not have any hint about the real sizes, zero seems to be the only safe return value.

Reply:
Done.

Reply:
@psychocoderHPC : Reporting zero here breaks PIConGPU via pmacc, which apparently checks this and then fails to allocate memory because there are only 0B left. How do you suggest to resolve this? Would you change this to be more optimistic (basically `INT64_MAX`) or should this be handled in pmacc?

Reply:
@BenjaminW3 Maybe there should be a trait added to check if the value returned by this one can be trusted on the given backend or should be ignored.

Reply:
Or return an optional

Reply:
Using optional would be a breaking change. With this:
```c++
template<class TDev>
ALPAKA_FN_HOST static auto getMemBytes(TDev const & dev)
 -> boost::optional<std::size_t>
{...}
```
All uses would need to change to:
```c++
auto memBytes = getMemBytes(dev);
if(memBytes)
{
 if(*memBytes < howMuchWeNeed)
// ...
}
else
{
// hope there is enough
}
```

Reply:
But better explicitly break the interface to make users aware than hiding the still breaking change by reporting UINT_MAX or similar unknown values.
Lets wait for an answer from @psychocoderHPC. I would still prefer returning zero.

Reply:
I am with you on the zero vs. UINT_MAX issue.

Adding another trait would however be less breaking than returning an optional: Using optional would require all code using any backend to change. Adding another trait, would allow legacy code work unchanged with all backends other than this one. Only code that wants to support this new backend would have to add a check with the new trait ( `deviceKnowsAvailableMem` or something ), which with time would become all code, but it could happen more gradually.

Reply:
Is it possible to start a kernel to figure out which type of device is used? Than we could use the return value of this kernel to select they way how to calculate the available memory.

Reply:
I do not see any standard way to get the type of device being used. We could add a cmake option to select the target which could then define a macro and set the appropriate compiler flags. Something like this should become part of the OpenMP cmake module at some point. This would, however, not work when the stand-alone headers are used.

Even if we could determine what type of device is being used, there would still be no guarantee, that we can determine a value, because we need code for each device type, so we would have nothing in a new environment.

We need a way to tell the program, that we cannot get this information as a fallback.

Reply:
Also: If this would be a runtime decision, all strategies for possible targets need to be compiled. While it is possible to leave out runtimes which are not available at compile time, this would make CUDA a dependency if you want to compile for OpenMP 5 targeting nvptx,for example. Usually this dependency would not be explicit and it might sometimes be hidden away in the compiler.

Reply:
I had a discussion with @psychocoderHPC 
* The PIConGPU/pmacc issue is not central and can be worked around very easily
* He brought up the option of raising an exception in the specialization, which would only break application if they use this backend and the root cause would be easy to track down. Also these functions are not pefrormance critical.
* He would be OK with just returning zero. I will add a commit to this PR documenting the possibility of this trait returning 0, which then means, that the real value is unknown. I check this issue for the OpenACC backend: There it is possible to get these values (works with NHVPC 20.7+GPU). However GCC just returns 0, which maybe because of a lack of support or disabled offloading, and a lack of support in the runtime for the host. This means, 0 return values can also occur for this trait when support is technically there.

Reply:
Added https://github.com/alpaka-group/alpaka/pull/1126/commits/092b458b133288b375f4e9abda977396d15b9b81

I suggest to move forward with this state and optionally create an issue in case we want to consider more invasive handling (optional, exception, ...).